llm : add Falcon support

ggerganov opened the pull request:
ref: #1602 

This PR adds support for Falcon models in `llama.cpp`

Currently, I've put everything inside `llama.cpp` source file.
We can think about better refactoring so that we can scale the process and add more models without the source file becoming too big. But for now, the main goal was to see what it takes to add support for a new LLM and serve as an example.

The PR also implements a more accurate BPE tokenizer utilizing merges. Used a reference implementation from:

https://github.com/cmp-nct/ggllm.cpp

However, I've dropped the unicode lib for clarity and therefore the implementation does not produce exactly correct tokenization in all cases. It should be good enough for latin languages. The advantage is that the code is more compact. Hopefully in the future we will improve it. In any case, 3rd party tokenizers are always an option and work well with `llama.cpp`, so if accuracy is essential - this is the recommended way. 

- [x] CPU
- [x] Metal
- [x] CUDA
- [ ] OpenCL (not tested)

**CUDA offloading still does not work due to missing RoPE NeoX kernel**

### TODO

- [x] GGUF typed access to KV and tensor names
- [x] refactor `llama_model_load_internal()` into reusable steps
- [x] implement `llm_load_falcon()` similar to `llm_load_llama()`
- [x] implement `llm_build_falcon()` similar to `llm_build_llama()`
- [x] add Metal support
- [x] add quantization support
- [x] GPU offloading for CUDA
- [x] RoPE (mode = 2) implementation for CUDA backends
- [x] Clean-up the `llm_build_falcon()` function
- [x] Fix tokenization
- [ ] The `bpe_gpt2_preprocess()` is quite slow (regex)

### Usage

```bash
# get HF model
git clone https://huggingface.co/tiiuae/falcon-40b

# convert to F16 .gguf
python3 convert-falcon-hf-to-gguf.py ./falcon-40b/ 1

# quantize to Q4_0
make -j && ./quantize ./falcon-40b/ggml-model-f16.gguf ./falcon-40b/ggml-model-q4_0.gguf q4_0

# run
make -j && ./main -m ./falcon-40b/ggml-model-q4_0.gguf
```

### Performance

- M2 Ultra

| model                          | backend    | n_gpu_layers | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------- | --------------: |
| Falcon 7B mostly F16  | Metal      |          999 | pp 512     |   400.84 ± 1.94 |
| Falcon 7B mostly Q8_0          | Metal      |          999 | pp 512     |   387.38 ± 1.31 |
| Falcon 7B mostly Q4_0          | Metal      |          999 | pp 512     |   388.54 ± 1.85 |
| Falcon 7B mostly Q4_1          | Metal      |          999 | pp 512     |   390.00 ± 1.07 |
| Falcon 7B mostly F16  | Metal      |          999 | tg 64      |    29.84 ± 0.04 |
| Falcon 7B mostly Q8_0          | Metal      |          999 | tg 64      |    61.96 ± 0.07 |
| Falcon 7B mostly Q4_0          | Metal      |          999 | tg 64      |    91.64 ± 0.16 |
| Falcon 7B mostly Q4_1          | Metal      |          999 | tg 64      |    84.04 ± 0.04 |
| Falcon 40B mostly F16          | Metal      |          999 | pp 512     |    78.00 ± 0.14 |
| Falcon 40B mostly Q8_0         | Metal      |          999 | pp 512     |    77.31 ± 0.13 |
| Falcon 40B mostly Q4_0         | Metal      |          999 | pp 512     |    77.43 ± 0.19 |
| Falcon 40B mostly Q4_1         | Metal      |          999 | pp 512     |    77.87 ± 0.08 |
| Falcon 40B mostly F16          | Metal      |          999 | tg 64      |     5.94 ± 0.00 |
| Falcon 40B mostly Q8_0         | Metal      |          999 | tg 64      |    13.78 ± 0.01 |
| Falcon 40B mostly Q4_0         | Metal      |          999 | tg 64      |    22.86 ± 0.02 |
| Falcon 40B mostly Q4_1         | Metal      |          999 | tg 64      |    21.26 ± 0.01 |

build: 38b16df (1052)

| model                          | backend    | n_gpu_layers | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------- | --------------: |
| LLaMA v2 7B mostly Q4_0        | Metal      |            1 | pp 512     |   632.43 ± 0.18 |
| LLaMA v2 13B mostly Q4_0       | Metal      |            1 | pp 512     |   369.02 ± 0.07 |
| LLaMA v2 70B mostly Q4_0       | Metal      |            1 | pp 512     |    76.41 ± 0.01 |
| LLaMA v2 7B mostly Q4_0        | Metal      |            1 | tg 128     |    86.13 ± 0.11 |
| LLaMA v2 13B mostly Q4_0       | Metal      |            1 | tg 128     |    54.52 ± 0.06 |
| LLaMA v2 70B mostly Q4_0       | Metal      |            1 | tg 128     |    14.39 ± 0.01 |

build: 176ea71 (1052)

- CUDA under WSL2:
  Device 0: NVIDIA GeForce RTX 3090 Ti, compute capability 8.6

| model                          | backend    | n_gpu_layers | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------- | --------------: |
| LLaMA 7B mostly Q4_0 (guessed) | CUDA       |           33 | pp 512     |  686.76 ± 15.63 |
| falcon-7b 7B mostly Q4_0       | CUDA       |           33 | pp 512     |   476.94 ± 1.95 |
| LLaMA 7B mostly Q4_0 (guessed) | CUDA       |           33 | tg 128     |    44.67 ± 0.28 |
| falcon-7b 7B mostly Q4_0       | CUDA       |           33 | tg 128     |    68.52 ± 0.20 |

| model                          | backend    | n_gpu_layers | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------- | --------------: |
| LLaMA 30B mostly Q4_0          | CUDA       |           99 | pp 512     |   615.45 ± 3.24 |
| falcon-40b 40B mostly Q4_0     | CUDA       |           99 | pp 512     |   441.00 ± 0.93 |
| LLaMA 30B mostly Q4_0          | CUDA       |           99 | tg 128     |    18.07 ± 0.03 |
| falcon-40b 40B mostly Q4_0     | CUDA       |           99 | tg 128     |    17.29 ± 0.01

build: 176ea71 (1052)

ggerganov committed:
llama : refactor GGUF constants into static maps

ggerganov committed:
llama : check if model architecture is known

ggerganov committed:
llama : refactor llama_model_load_internal()

ggerganov committed:
gguf : add KV constant maps

ggerganov committed:
Merge branch 'master' into falcon

ggerganov committed:
llm : read arch-specific KVs

ggerganov committed:
convert : add dummy scores + types

ggerganov committed:
falcon : load tensor data (CPU only)

ggerganov committed:
llama : fix loading progress bar

ggerganov committed:
llama : add arch member to llama_model

ggerganov committed:
falcon : CPU inference working

klosax commented:
Conversion of 7b model do not work. The qkv transform needs n_kv_head = 1 for this to work
https://github.com/ggerganov/llama.cpp/blob/3c7c325b9867a30637529b0328fbc73b4e527004/convert-falcon-hf-to-gguf.py#L204
https://github.com/ggerganov/llama.cpp/blob/3c7c325b9867a30637529b0328fbc73b4e527004/convert-falcon-hf-to-gguf.py#L241-L246

ggerganov committed:
falcon : support non-40B models

klosax commented:
main using falcon-7b:
`Error loading model: create_tensor: tensor 'blk.0.attn_norm_2.weight' not found`

ggerganov committed:
falcon : minor

ggerganov commented:
@klosax Reconvert and it works

Strangely, enabling Metal we crash here:

https://github.com/ggerganov/llama.cpp/blob/0ec27ad66c56a4831f62a8106b7873a6818bf051/llama.cpp#L2679

If I disable the graph concurrency optimization, it does not crash:

https://github.com/ggerganov/llama.cpp/blob/0ec27ad66c56a4831f62a8106b7873a6818bf051/llama.cpp#L4799-L4800

Any guesses what could be wrong?

klosax commented:
> Any guesses what could be wrong?
Not a clue. I can try enabling cublas to see if that works.

ggerganov committed:
llama : minor updates

ggml-ci

klosax committed:
convert-falcon-hf-to-gguf.py : fix special token mapping

klosax committed:
llama.cpp : llama default UNK token = id 0

klosax commented:
Perplexity on falcon-7b-q4_0 wont work:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Aborted (core dumped)
```

ggerganov commented:
Yes, something wrong with the tokenization. Here is the stack trace:

<details>

```java
system_info: n_threads = 16 / 24 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
libc++abi: terminating due to uncaught exception of type std::out_of_range: unordered_map::at: key not found
Process 26597 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x000000018b3d8764 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`:
->  0x18b3d8764 <+8>:  b.lo   0x18b3d8784               ; <+40>
    0x18b3d8768 <+12>: pacibsp 
    0x18b3d876c <+16>: stp    x29, x30, [sp, #-0x10]!
    0x18b3d8770 <+20>: mov    x29, sp
Target 0: (perplexity) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x000000018b3d8764 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000018b40fc28 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x000000018b31dae8 libsystem_c.dylib`abort + 180
    frame #3: 0x000000018b3c8b84 libc++abi.dylib`abort_message + 132
    frame #4: 0x000000018b3b83b4 libc++abi.dylib`demangling_terminate_handler() + 320
    frame #5: 0x000000018b08f03c libobjc.A.dylib`_objc_terminate() + 160
    frame #6: 0x000000018b3c7f48 libc++abi.dylib`std::__terminate(void (*)()) + 16
    frame #7: 0x000000018b3cad34 libc++abi.dylib`__cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) + 36
    frame #8: 0x000000018b3cace0 libc++abi.dylib`__cxa_throw + 140
    frame #9: 0x0000000100026ba4 perplexity`std::__1::__throw_out_of_range[abi:v15006](__msg="unordered_map::at: key not found") at stdexcept:268:5
    frame #10: 0x00000001000619bc perplexity`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, int> > >::at(this=0x00000001006040e0 size=65023, __k="<0xE7>") const at unordered_map:1863:9
    frame #11: 0x00000001000615a8 perplexity`llama_byte_to_token(vocab=0x00000001006040d8, ch='\xe7') at llama.cpp:2852:30
    frame #12: 0x000000010005cb04 perplexity`llama_tokenizer::resegment(this=0x000000016fdfdf08, symbol=0x0000000130fe5888, output=size=1581) at llama.cpp:2991:44
    frame #13: 0x000000010005be4c perplexity`llama_tokenizer::tokenize(this=0x000000016fdfdf08, text=" \n = Robert Boulter = \n \n Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starring role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons written by Simon Stephens , which was performed in 2001 at the Royal Court Theatre . He had a guest role in the television series Judge John Deed in 2002 . In 2004 Boulter landed a role as \" Craig \" in the episode \" Teddy 's Story \" of the television series The Long Firm ; he starred alongside actors Mark Strong and Derek Jacobi . He was cast in the 2005 theatre productions of the Philip Ridley play Mercury Fur , which was performed at the Drum Theatre in Plymouth and the Menier Chocolate Factory in London . He was directed by John Tiffany and starred alongside Ben Whishaw , Shane Zaza , Harry Kent , Fraser Ayres , Sophie Stanton and Dominic Hall . \n In 2006 , Boulter starred alongside Whishaw in the play Citizenship written by Mark Ravenhill . He appeared on a 2006 episode of the televi"..., output=size=1581) at llama.cpp:2971:13
    frame #14: 0x0000000100036bd8 perplexity`llama_tokenize_internal(vocab=0x00000001006040d8, raw_text=" \n = Robert Boulter = \n \n Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starring role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons written by Simon Stephens , which was performed in 2001 at the Royal Court Theatre . He had a guest role in the television series Judge John Deed in 2002 . In 2004 Boulter landed a role as \" Craig \" in the episode \" Teddy 's Story \" of the television series The Long Firm ; he starred alongside actors Mark Strong and Derek Jacobi . He was cast in the 2005 theatre productions of the Philip Ridley play Mercury Fur , which was performed at the Drum Theatre in Plymouth and the Menier Chocolate Factory in London . He was directed by John Tiffany and starred alongside Ben Whishaw , Shane Zaza , Harry Kent , Fraser Ayres , Sophie Stanton and Dominic Hall . \n In 2006 , Boulter starred alongside Whishaw in the play Citizenship written by Mark Ravenhill . He appeared on a 2006 episode of the televi"..., bos=true, escape=false) at llama.cpp:3055:15
    frame #15: 0x00000001000367b0 perplexity`::llama_tokenize_with_model(model=0x0000000100604080, text=" \n = Robert Boulter = \n \n Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starring role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons written by Simon Stephens , which was performed in 2001 at the Royal Court Theatre . He had a guest role in the television series Judge John Deed in 2002 . In 2004 Boulter landed a role as \" Craig \" in the episode \" Teddy 's Story \" of the television series The Long Firm ; he starred alongside actors Mark Strong and Derek Jacobi . He was cast in the 2005 theatre productions of the Philip Ridley play Mercury Fur , which was performed at the Drum Theatre in Plymouth and the Menier Chocolate Factory in London . He was directed by John Tiffany and starred alongside Ben Whishaw , Shane Zaza , Harry Kent , Fraser Ayres , Sophie Stanton and Dominic Hall . \n In 2006 , Boulter starred alongside Whishaw in the play Citizenship written by Mark Ravenhill . He appeared on a 2006 episode of the televi"..., tokens=0x00000001187a0000, n_max_tokens=1290590, add_bos=true) at llama.cpp:5480:16
    frame #16: 0x000000010003671c perplexity`::llama_tokenize(ctx=0x0000000102008200, text=" \n = Robert Boulter = \n \n Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starring role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons written by Simon Stephens , which was performed in 2001 at the Royal Court Theatre . He had a guest role in the television series Judge John Deed in 2002 . In 2004 Boulter landed a role as \" Craig \" in the episode \" Teddy 's Story \" of the television series The Long Firm ; he starred alongside actors Mark Strong and Derek Jacobi . He was cast in the 2005 theatre productions of the Philip Ridley play Mercury Fur , which was performed at the Drum Theatre in Plymouth and the Menier Chocolate Factory in London . He was directed by John Tiffany and starred alongside Ben Whishaw , Shane Zaza , Harry Kent , Fraser Ayres , Sophie Stanton and Dominic Hall . \n In 2006 , Boulter starred alongside Whishaw in the play Citizenship written by Mark Ravenhill . He appeared on a 2006 episode of the televi"..., tokens=0x00000001187a0000, n_max_tokens=1290590, add_bos=true) at llama.cpp:5450:12
    frame #17: 0x0000000100014f50 perplexity`llama_tokenize(ctx=0x0000000102008200, text=" \n = Robert Boulter = \n \n Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starring role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons written by Simon Stephens , which was performed in 2001 at the Royal Court Theatre . He had a guest role in the television series Judge John Deed in 2002 . In 2004 Boulter landed a role as \" Craig \" in the episode \" Teddy 's Story \" of the television series The Long Firm ; he starred alongside actors Mark Strong and Derek Jacobi . He was cast in the 2005 theatre productions of the Philip Ridley play Mercury Fur , which was performed at the Drum Theatre in Plymouth and the Menier Chocolate Factory in London . He was directed by John Tiffany and starred alongside Ben Whishaw , Shane Zaza , Harry Kent , Fraser Ayres , Sophie Stanton and Dominic Hall . \n In 2006 , Boulter starred alongside Whishaw in the play Citizenship written by Mark Ravenhill . He appeared on a 2006 episode of the televi"..., add_bos=true) at common.cpp:711:16
    frame #18: 0x0000000100003284 perplexity`perplexity(ctx=0x0000000102008200, params=0x000000016fdfedd8) at perplexity.cpp:35:19
    frame #19: 0x0000000100006034 perplexity`main(argc=8, argv=0x000000016fdff258) at perplexity.cpp:412:9
    frame #20: 0x000000018b0b7f28 dyld`start + 2236
(lldb) frame select 11
frame #11: 0x00000001000615a8 perplexity`llama_byte_to_token(vocab=0x00000001006040d8, ch='\xe7') at llama.cpp:2852:30
   2849	   char buf[7];
   2850	   int result = snprintf(buf, sizeof(buf), "<0x%02X>", ch);
   2851	   GGML_ASSERT(0 <= result && result < 7);
-> 2852	   return vocab.token_to_id.at(buf);
   2853	}
   2854	
   2855	static std::string llama_escape_whitespace(const std::string& text) {
(lldb) print ch
(uint8_t) $0 = '\xe7'
(lldb) 
```
</details>

goerch commented:
> Yes, something wrong with the tokenization. Here is the stack trace:

Is `Falcon` using a `sentencepiece`-based tokenizer?

klosax commented:
> Is `Falcon` using a `sentencepiece`-based tokenizer?

No, it is bpe.

goerch commented:
> No, it is bpe.

So I'd recommend switching [vocabulary type](https://github.com/ggerganov/llama.cpp/blob/c63bb1d16a70c03440671b76954bb767513cead8/llama.cpp#L1304-L1311). That will be a first for me ;)

klosax commented:
`‎llama_tokenize_bpe‎()` is never used in llama.cpp ?

klosax commented:
> Edit: is there an easy way for me to test this?

Build the branch and test with Falcon-7b from here:
https://huggingface.co/tiiuae/falcon-7b/tree/main

slaren commented:
Seeing that falcon is entirely in bfloat16, should it be converted as f32?

goerch commented:
> `‎llama_tokenize_bpe‎()` and `llama_token_to_str_bpe‎()` is never used in llama.cpp ?

These are used in test to assert low level behavior of the tokenizer.

klosax commented:
> Seeing that falcon is entirely in bfloat16, should it be converted as f32?

It works using simpler dataset than wiki.test.raw It is a tokenizer problem.

And the bfloat16 is converted to float32 in the conversion script.

klosax commented:
> These are used in test to assert low level behavior of the tokenizer.

Ok.

slaren commented:
> And the bfloat16 is converted to float32 in the conversion script.

Maybe I am misunderstanding, but the output from `convert-falcon-hf-to-gguf.py` when converting as f16 suggests that bfloat16 is converted to float16. The possible issue is that the range of float16 is lower than bfloat16, but maybe the actual values of the tensors are within the range of float16, or it is just not a problem in practice.
```
token_embd.weight, n_dims = 2, torch.bfloat16 --> float16
blk.0.attn_norm.weight, n_dims = 1, torch.bfloat16 --> float32
blk.0.attn_norm.bias, n_dims = 1, torch.bfloat16 --> float32
blk.0.attn_qkv.weight, n_dims = 2, torch.bfloat16 --> float16
blk.0.attn_output.weight, n_dims = 2, torch.bfloat16 --> float16
blk.0.ffn_up.weight, n_dims = 2, torch.bfloat16 --> float16
blk.0.ffn_down.weight, n_dims = 2, torch.bfloat16 --> float16
```

klosax commented:
Perplexity falcon-7b-q4_0 on [tinystories-valid.txt.1000](https://github.com/klosax/misc/blob/main/tinystories-valid.txt.1000)
[78]4.2808

klosax commented:
> The possible issue is that the range of float16 is lower than bfloat16, but maybe the actual values of the tensors are within the range of float16, or it is just not a problem in practice.

I guess converting bf16 to f16 will cut some outliers. The original LLaMA PTH models are also bf6, which have a very slightly lower perplexity than the f16.

klosax commented:
Here is `wiki.test.raw.406` which get the tokenizer errors:
https://github.com/klosax/misc/blob/main/wiki.test.raw.406

slaren commented:
The crash in the tokenizer is here:
```
#12 0x00005555555b5a12 in llama_byte_to_token (vocab=..., ch=231 '\347') at llama.cpp:2852
#13 0x00005555555cc9d5 in llama_tokenizer::resegment (this=0x7fffffffc3d0, symbol=..., output=std::vector of length 1581, capacity 2048 = {...}) at llama.cpp:2991
#14 0x00005555555cc84b in llama_tokenizer::tokenize (this=0x7fffffffc3d0,
```

goerch commented:
> So I guess that some bytes don't have a token?

I`ve seen this with `Aquila` before. I believe we might have to send this through the 'real' BPE tokenizer then?

klosax commented:
> : I've seen this with `Aquila` before. I believe we might have to send this through the 'real' BPE tokenizer then?

But what is the problem exactly?

klosax commented:
I guess the bpe needs to use the merges?

slaren commented:
I am not sure what is the issue, but this seems to work:
```C++
static llama_token llama_byte_to_token(const llama_vocab & vocab, uint8_t ch) {
    return vocab.token_to_id.at(std::string(1, ch));
}
```

klosax commented:
That looks like something sentencepiece?
https://github.com/ggerganov/llama.cpp/blob/ffa5099c6d19bf9e783e545ecd3040b75d06c06c/llama.cpp#L2848-L2853

goerch commented:
> I am not sure what is the issue, but this seems to work:

> That looks like something sentencepiece?

If you are going to differentiate the vocab type there this could work :)

klosax commented:
> I am not sure what is the issue, but this seems to work:
ppl wiki.text.raw falcon 7b q4_0:
[1]11.4635,[2]16.0172,[3]15.6642,^C

klosax commented:
> Edit: still fighting with git :(

you could also download the branch as zip

klosax commented:
> If you are going to differentiate the vocab type there this could work :)

This works on both llama and falcon

```
    if (vocab.type == LLAMA_VOCAB_TYPE_SPM) {
        char buf[7];
        int result = snprintf(buf, sizeof(buf), "<0x%02X>", ch);
        GGML_ASSERT(0 <= result && result < 7);
        return vocab.token_to_id.at(buf);
    }
    // vocab.type == LLAMA_VOCAB_TYPE_BPE
    return vocab.token_to_id.at(std::string(1, ch));
```

klosax committed:
llama.cpp : fix bpe tokenizer

klosax committed:
llama.cpp : fix the fix of bpe tokenizer

klosax commented:
Running ppl on `wiki.text.raw.406` both using this simple bpe tokenizer and using the advanced tokenizer in the gptneox examples.

klosax commented:
Perplexity using simple llama.cpp bpe tokenizer vs advanced tokenizer

Falcon-7b

wiki.test.raw.406

q4_0-simple: 12.8792
q4_0-advanced: 8.4978

WilliamTambellini commented:
Cool. 
What about lora adapter support? 
As today convert-lora-to-ggml.py seems hard coded to search/expect llama tensors and hits on falcon tensors:
```
Error: unrecognized tensor base_model.model.transformer.h.0.self_attention.query_key_value.lora_A.weight
``` 
the regex is
https://github.com/ggerganov/llama.cpp/blob/3b6cfe7c927df178ca3c11643c3ec93e143471c9/convert-lora-to-ggml.py#L29
```
def translate_tensor_name(t: str) -> str:
    match = re.match(r".*layers\.(\d+)\.(\w+\.\w+)\.lora_(A|B)\.weight", t)
```
Best

lshzh-ww commented:
@ggerganov 
> Strangely, enabling Metal we crash here:
> If I disable the graph concurrency optimization, it does not crash:
> Any guesses what could be wrong?

This is because the computation graph of Falcon changes between prompt evaluation and inference. When `n_tokens=1`, we omit all the `GGML_REPEAT` nodes in the graph. However, they naturally appear when `n_tokens>1`. This violates our previous assumption that the topology of the graph doesn't change.

I suggest that for `ggml_repeat`, instead of simply returning `src` when `dst` is of the same shape, we still create an empty `GGML_OP_NONE` node. This ensures that the shape of the computation graph remains consistent. Avoiding changes to the computation graph is also useful if we intend to perform any further graph optimizations in the future.

ggerganov commented:
@slaren 

> > And the bfloat16 is converted to float32 in the conversion script.
> 
> Maybe I am misunderstanding, but the output from `convert-falcon-hf-to-gguf.py` when converting as f16 suggests that bfloat16 is converted to float16. The possible issue is that the range of float16 is lower than bfloat16, but maybe the actual values of the tensors are within the range of float16, or it is just not a problem in practice.
> 
> ```
> token_embd.weight, n_dims = 2, torch.bfloat16 --> float16
> blk.0.attn_norm.weight, n_dims = 1, torch.bfloat16 --> float32
> blk.0.attn_norm.bias, n_dims = 1, torch.bfloat16 --> float32
> blk.0.attn_qkv.weight, n_dims = 2, torch.bfloat16 --> float16
> blk.0.attn_output.weight, n_dims = 2, torch.bfloat16 --> float16
> blk.0.ffn_up.weight, n_dims = 2, torch.bfloat16 --> float16
> blk.0.ffn_down.weight, n_dims = 2, torch.bfloat16 --> float16
> ```

I hope that the induced error is not too big, but maybe a more thorough study would be useful.
I guess if the user needs F16 precision, there is no way around this without adding BF16 support to `ggml`. When generating quantized models, they can convert first to F32 as intermediate step and use that to quantize for maximum precission.

ggerganov committed:
ggml : pass eps to ggml_norm

ggerganov committed:
metal : implement RoPE (mode = 2) + avoid ggml_repeat

ggerganov committed:
ggml : ggml_repeat always creates new tensor

ggerganov commented:
@lshzh-ww 

Thanks for the help - I've changed `ggml_repeat` to always return a tensor regardless of the input shapes.
Additionally, I've removed `ggml_repeat` from Falcon graph and rely on broadcasts. The Metal inference now does not crash, but there is still some bug because it outputs junk. Will look into fixing this now

ggerganov committed:
falcon : copy-paste self-attention from LLaMA

klosax commented:
> How do you compute the "advanced" perplexity values?

I'm using a tokenizer capable of handling the merges section of the vocab. The one  that the simple falcon / gptneox example uses.

ggerganov committed:
metal : print extra compute pipeline info

klosax commented:
> I hope that the induced error is not too big, but maybe a more thorough study would be useful.

I will check ppl on F32 vs F16.

ggerganov commented:
While searching for the Metal bug, I discovered that each kernel has a value called `maxTotalThreadsPerThreadgroup`:

https://developer.apple.com/documentation/metal/mtlcomputepipelinestate/1414927-maxtotalthreadsperthreadgroup?language=objc

We should start respecting this value to get maximum efficiency.

```java
ggml_metal_init: allocating
ggml_metal_init: loading '/Users/ggerganov/development/github/llama.cpp/ggml-metal.metal'
ggml_metal_init: loaded kernel_add                            0x104508490 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_add_row                        0x104508bd0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul                            0x104509110 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_row                        0x104509760 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_scale                          0x104509ca0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_silu                           0x10450a1e0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_relu                           0x10450a720 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_gelu                           0x10450ac60 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_soft_max                       0x10450b330 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_diag_mask_inf                  0x10450b9b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_f16                   0x10450c080 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_0                  0x10450c8c0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_1                  0x10450cf90 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q2_K                  0x10450d660 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q3_K                  0x10450dd30 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_K                  0x10450e400 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q5_K                  0x10450ead0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q6_K                  0x10450f1a0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_rms_norm                       0x10450f880 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_norm                           0x1045100c0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_f16_f32                0x104510990 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_0_f32               0x104511110 | th_max =  896 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_1_f32               0x104511890 | th_max =  896 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q2_K_f32               0x104512190 | th_max =  640 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q3_K_f32               0x104512910 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_K_f32               0x104513090 | th_max =  576 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q5_K_f32               0x104513810 | th_max =  576 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q6_K_f32               0x104514190 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_f16_f32                 0x104514bb0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_0_f32                0x104515370 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_1_f32                0x104515b30 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q2_K_f32                0x1045162f0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q3_K_f32                0x104516830 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_K_f32                0x104516ff0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q5_K_f32                0x1045177b0 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q6_K_f32                0x104517f70 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_rope                           0x1045184b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_alibi_f32                      0x104518d90 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f32_f16                    0x104519640 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f32_f32                    0x104519ef0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f16_f16                    0x10451a7a0 | th_max = 1024 | th_width =   32
ggml_metal_init: recommendedMaxWorkingSetSize  = 147456.00 MB
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: maxTransferRate               = built-in GPU
```

klosax commented:
Perplexity `wiki.text.raw.406`

All tests without BOS.

Using the simple built-in bpe tokenizer:

F32: 10.6728
F16: 10.6729
Q8_0: 10.6886
Q4_0: 12.8732

Using bpe tokenizer utilizing merges:

F32: 7.20140
F16: 7.20136
Q8_0: 7.2116
Q4_0: 8.4978

ggerganov committed:
falcon : minor changes (still chasing the Metal problem)

klosax commented:
Since comparing ppl of different models is not a good idea due to the differences in training datasets, I made some tests measuring the HellaSwag accuracy:

400 tasks on Falcon-7b-Q8_0

Simple built-in bpe tokenizer: 66.50 %
Bpe tokenizer utilizing merges: 76.50 %

For comparison:

OpenLLaMA-v2-3b: 69.75 %
LLaMA-v1-7b: 75.25 %
LLaMA-v2-7b: 77.50 %
More at https://github.com/ggerganov/llama.cpp/discussions/2321

klosax committed:
llama.cpp : fix linefeed token

ggerganov committed:
metal : fix GELU kernel numerical stability by using precise::tanh

ggerganov commented:
Metal inference now works correctly without the graph concurrency optimization. The bug was related to `tanh` returning NaNs sometimes:

https://github.com/ggerganov/llama.cpp/blob/0a85ae73977e69b0fbeb9345361ce8e358919416/ggml-metal.metal#L85-L96

Will try to fix the concurrency optimization now. I think it somehow gets confused due to the attention and feed-forward section being completely parallel in Falcon

ggerganov committed:
metal : temporary workaround for the concurrency optimization bug

slaren committed:
falcon : add CUDA offloading (#2739)

ggerganov committed:
llama : better model naming and size reporting

ggerganov commented:
Here are some performance stats for Falcon on M2 Ultra:

| model                          | backend    | n_gpu_layers | test       |             t/s |
| ------------------------------ | ---------- | -----------: | ---------- | --------------: |
| LLaMA v2 7B mostly Q4_0        | Metal      |            1 | pp 512     |   632.43 ± 0.18 |
| LLaMA v2 13B mostly Q4_0       | Metal      |            1 | pp 512     |   369.02 ± 0.07 |
| LLaMA v2 70B mostly Q4_0       | Metal      |            1 | pp 512     |    76.41 ± 0.01 |
| Falcon 7B mostly Q4_0          | Metal      |            1 | pp 512     |   390.20 ± 1.28 |
| Falcon 40B mostly Q4_0         | Metal      |            1 | pp 512     |    77.97 ± 0.05 |
| LLaMA v2 7B mostly Q4_0        | Metal      |            1 | tg 128     |    86.13 ± 0.11 |
| LLaMA v2 13B mostly Q4_0       | Metal      |            1 | tg 128     |    54.52 ± 0.06 |
| LLaMA v2 70B mostly Q4_0       | Metal      |            1 | tg 128     |    14.39 ± 0.01 |
| Falcon 7B mostly Q4_0          | Metal      |            1 | tg 128     |    84.08 ± 0.10 |
| Falcon 40B mostly Q4_0         | Metal      |            1 | tg 128     |    21.73 ± 0.01 |

build: 176ea71 (1052)

ggerganov commented:
Btw, I believe it is using the wrong rope function. We need to add support for `mode == 2`.
Here is the Metal code:

https://github.com/ggerganov/llama.cpp/blob/176ea716b355cff5a2b6a97c7648cee138183818/ggml-metal.metal#L561-L598

ggerganov commented:
@klosax 

The perplexity results using the "simple" BPE tokenizer are significantly worse than the "advanced". I was hoping the differences to be smaller.

Looks like we will need the advanced BPE tokenizer added to `llama.cpp` after all.

I looked at the example implementation that you provided - this will be very helpful to use while implementing.
I wish the code was simpler so we can put it straight into `llama.cpp`, but it's not the case, so it will have to live in a separate C++ header file for now.

Will probably implement this after we merge the current PR

Should I change this string to "falcon" or keep it "gpt2"?

https://github.com/ggerganov/llama.cpp/blob/176ea716b355cff5a2b6a97c7648cee138183818/convert-falcon-hf-to-gguf.py#L122

Also, are there other models that use this specific tokenizer? I assume most GPT-NeoX-based models (e.g. Dolly, MPT) should be compatible, correct?

ggerganov committed:
Merge branch 'master' into falcon

klosax commented:
> Should I change this string to "falcon" or keep it "gpt2"?

No keep `gpt2` since it is the original gpt2 tokenizer model. The LLaMA models uses a special version of sentencepiecce SPM that is why it should be named `llama`, replit model uses another version of SPM

> Also, are there other models that use this specific tokenizer? I assume most GPT-NeoX-based models (e.g. Dolly, MPT) should be compatible, correct?

Yes most models besides LLaMA use the gpt2 bpe tokenizer.

ggerganov committed:
llama : prep new tokenizer support

ggerganov committed:
llama : advanced BPE tokenizer based on ggllm.cpp imlpementation

ggerganov commented:
@klosax 

Just pushed a new BPE tokenizer in `llama.cpp` - the perplexity is much better now.
I did take some shortcuts, like avoiding the Unicode lib, so most likely the implementation is not 100% correct, but I hope it is good enough. I like it this way because it fits neatly in `llama.cpp`

The only thing that is left is to implement CUDA RoPE and I will merge this.
Will give this a try, but if it turns too difficult to figure out - will leave for someone else to do it

ggerganov committed:
llama : remove oboslete comment

ggml-ci

ggerganov committed:
common : remove obsolete BPE API + disable test-tokenizer-1

ggerganov committed:
llama : revert BPE special-case in llama_byte_to_token()

klosax commented:
> Just pushed a new BPE tokenizer in `llama.cpp` based on your example - the perplexity is much better now.

Yes, it seems to work good enough now. I will run some tests and post the results.

> I did take some shortcuts, like avoiding the Unicode lib, so most likely the implementation is not 100% correct,

I guess the Unicode stuff is needed for non-latin languages like chinese.

klosax commented:
The default special tokens should be arch specific

Falcon
BOS id = 11
EOS id = 11
UNK none, id = -1

ggerganov committed:
cuda : add TODOs for RoPE NeoX implementation

klosax commented:
Falcon-7b

Current bpe tokenizer utiliizing merges:

quant | ppl
 :-- | --: 
F32 | 7.2360
F16 | 7.2360
Q8_0 | 7.2704
Q4_0 | 8.6976

Unicode bpe tokenizer utilizing merges:

quant | ppl
 :-- | --: 
F32 | 7.2014
F16 | 7.2014
Q8_0 | 7.2116
Q4_0 | 8.4978

ggerganov committed:
llama : default special tokens based on vocab type

ggerganov commented:
Ok, looks good enough to me. I tried implementing the CUDA RoPE kernel, but I get a bit lost, so will leave this for later - hopefully we get some help with it.

Will merge this in a bit

klosax commented:
Falcon-7b seems sensitive to quantization: HellaSwag on Q4_0 is only 29.25 % !
Q8_0 seems to work good (77%).
Possibly some tensors that should not be quantized at all.
Testing other formats.

ggerganov committed:
perplexity : add log for start of tokenization

ggerganov commented:
Yes, currently we don't have proper quantization mixtures as with LLaMA models. When you use `Q4_0` with Falcon 7B, the output tensor is quantized to `Q4_0` due to shape restrictions (not divisible by 256) instead of `Q6_K` as with LLaMA. Apparently, this has a big impact on the results. We can improve later

klosax commented:
I believe the unicode implementation did not use regex for that reason?
https://github.com/ggerganov/llama.cpp/blob/fae8faa135942918a7c9ecc8c2fc26be7f61140d/llama.cpp#L3324

ggerganov commented:
Since you have a repro, can you try if the NaNs disappear after changing `tanhf` -> `tanh` in the following places:

https://github.com/ggerganov/ggml/blob/6319ae9ad7bdf9f834b2855d7e9fa70508e82f57/src/ggml.c#L3553

https://github.com/ggerganov/ggml/blob/6319ae9ad7bdf9f834b2855d7e9fa70508e82f57/src/ggml.c#L3562

klosax commented:
> Since you have a repro, can you try if the NaNs disappear after changing `tanhf` -> `tanh` in the following places:

That seems to work. I will run some more tests.

cmp-nct commented:
> @klosax
> 
> The perplexity results using the "simple" BPE tokenizer are significantly worse than the "advanced". I was hoping the differences to be smaller.
> 
> Looks like we will need the advanced BPE tokenizer added to `llama.cpp` after all.
> 
> I looked at the example implementation that you provided - this will be very helpful to use while implementing. I wish the code was simpler so we can put it straight into `llama.cpp`, but it's not the case, so it will have to live in a separate C++ header file for now.
> 
> Will probably implement this after we merge the current PR
> 
> Should I change this string to "falcon" or keep it "gpt2"?
> 
> https://github.com/ggerganov/llama.cpp/blob/176ea716b355cff5a2b6a97c7648cee138183818/convert-falcon-hf-to-gguf.py#L122
> 
> Also, are there other models that use this specific tokenizer? I assume most GPT-NeoX-based models (e.g. Dolly, MPT) should be compatible, correct?

I created that tokenizer when I realized that without the bpe "merges" support barely any longer token is correctly tokenized, including the special tokens which frequently are used in fine tunes also any european language requires it.

@klosax I did not use regex only because I did not like adding it as a huge dependency and I thought it's going to be very slow, in the end that decision cost me quite some time. Same reasoning for creating the custom unicode c++ library, just in that case I'm sure it was the right decision.

klosax commented:
> I created that tokenizer when I realized that without the bpe "merges" support barely any longer token is correctly tokenized, including the special tokens which frequently are used in fine tunes also any european language requires it.

Yes, the merges are important and adding the merges lowered the perplexity by 33%. The value of having merges in bpe is comparable to having the scores in sentencepiece.

> I did not use regex only because I did not like adding it as a huge dependency and I thought it's going to be very slow, in the end that decision cost me quite some time.

The current llama.cpp implementation of the tokenizer uses regex for simplicity and it is slow. I guess it will be replaced later.

> Same reasoning for creating the custom unicode c++ library, just in that case I'm sure it was the right decision.

Yes, your unicode library is a much better choice than depending on the huge ICU for full unicode support. If this is implemented in llama.cpp it could possibly also be used by the LLaMA sentencepiece tokenizer.

The importance of a good tokenizer should not be underestimated when it comes to generation quality.

Green-Sky commented:
o.o 360GB safetensors ......
The architecture in the model card looks like the 40B one (but scaled up). so it *should* be compatible. would be cool if someone with a large enough RAM system could try this.

Green-Sky commented:
You can try to create the GGUF yourself using the https://github.com/ggerganov/llama.cpp/blob/master/convert-falcon-hf-to-gguf.py

logicchains commented:
The normal `convert.py` fails with:

>  Traceback (most recent call last):
>   File "llama.cpp/./convert.py", line 1224, in <module>
>     main()
>   File "llama.cpp/./convert.py", line 1165, in main
>     model_plus = load_some_model(args.model)
>   File "llama.cpp/./convert.py", line 1087, in load_some_model
>     model_plus = merge_multifile_models(models_plus)
>   File "llama.cpp/./convert.py", line 620, in merge_multifile_models
>     model = merge_sharded([mp.model for mp in models_plus])
>   File "llama.cpp/./convert.py", line 599, in merge_sharded
>     return {name: convert(name) for name in names}
>   File "llama.cpp/./convert.py", line 599, in <dictcomp>
>     return {name: convert(name) for name in names}
>   File "llama.cpp/./convert.py", line 574, in convert
>     lazy_tensors: list[LazyTensor] = [model[name] for model in models]
>   File "llama.cpp/./convert.py", line 574, in <listcomp>
>     lazy_tensors: list[LazyTensor] = [model[name] for model in models]
> KeyError: 'transformer.h.0.mlp.dense_h_to_4h.weight'

logicchains commented:
Looks like convert.py expects all the `model-*-of-00081.safetensors` files to contain the same tensors, but actually for this model they contain different tensors.

logicchains commented:
Seems there is code in that for processing different tensors split across files, but changing it to that still fails, due to the config.json missing `intermediate_size` and `rms_norm_eps`.

logicchains commented:
If setting `n_ff` to `hidden_size * 4` like in `convert-falcon-hf-to-gguf`, the hard-coded logic in ` Params.find_n_mult` ends up failing: `failed to find n_mult for (n_ff=59392, n_embd=14848)`. If we just hard-code n_mult to 1, then the code ends up failing later when looking for the tokenizer, which I guess normal `convert.py` doesn't support.

Green-Sky commented:
> Seems there is code in that for processing different tensors split across files, but changing it to that still fails, due to the config.json missing `intermediate_size` and `rms_norm_eps`.

the config.json specifies:
```
"layer_norm_epsilon": 1e-05,
```

> If setting `n_ff` to `hidden_size * 4` like in `convert-falcon-hf-to-gguf`

iirc this should be something differently.

> the hard-coded logic in  `Params.find_n_mult` ends up failing: `failed to find n_mult for (n_ff=59392, n_embd=14848)`

should be save to ignore `n_mult` :+1:

> then the code ends up failing later when looking for the tokenizer, which I guess normal `convert.py` doesn't support.

yea, I *think* it is the same as falcon 40B

TheBloke commented:
I've been told that the model architecture should be identical to 40B, and I'm making GPTQs right now and it seems to work fine with the same code as worked with Falcon 40B.

I'd assumed convert-falcon-to-gguf.py would be the one to update, rather than convert.py?

I had a quick look at it myself earlier. There's a few easy fixes at the top, changing field names to match their new names in config.json.  But then it's written to load from pytorch_model.bin and Falcon 180 is in Safetensors and when I looked at the convert.py code for safetensors it does memory mapping and I gave up at that point.

If anyone can get a working script out I can have GGUFs up soon after that

logicchains commented:
I got a version of convert-falcon-to-gguf.py working: https://github.com/logicchains/llama.cpp/blob/falcon180B/convert-falcon180-hf-to-gguf.py (forked because I can't make a branch here). And it seems to just work! Not too slow either; 0.8 tokens/second for the 6bit quantisation, while with llama2 70b 8bit I get around 1.5 tokens/seconds.

 ./main -m ./models/falcon-180B-q6_K.gguf -c 2048 --temp 0.7 -t 32 -p "The secrets to a happy marriage are as follows:"
 
 
> 
>  The secrets to a happy marriage are as follows:
> - Communicate.
> - Be honest.
> - Do not be afraid to fight.
> - Have s\*x.
> Those are the basics. Those are the things that will keep your marriage healthy. Those are things you should try to do as much as possible. But those things aren’t magic. You can be doing all of those things, and your marriage can still fail. If you want the secret sauce for a happy marriage — the one thing that most people forget to do — then keep reading.
> Here’s what I know for sure: A happy marriage is not a marriage full of happiness. It’s not a marriage that lacks sadness, or pain, or anger, or frustration. A happy marriage is not a marriage without hard days and hard times. It’s not a marriage without arguments, or tears, or long nights where you lay in bed thinking, What have I done?
> A happy marriage is a marriage that is full of growth. It’s a marriage that has two people who are willing to be vulnerable, to open themselves up, and to say the hard things. A happy marriage is a marriage with two people who are both willing to change for the better. It’s a marriage with two people who are both willing to face their fears, their anxieties, and their demons head-on, no matter how scary it can be.
> A happy marriage is a marriage that is full of trust. It’s a marriage that has two people who are willing to share their deepest thoughts, their darkest secrets, and their most intimate moments. A happy marriage is a marriage where you feel safe enough to open yourself up completely.
> So what do you do when you want a happy marriage? You start by communicating. You start by being honest. You start by not being afraid to fight. And yes, of course, you have s\*x. But more than anything else, you have to be willing to grow together.
> It’s not easy, and it’s not always fun, but it’s worth it. Trust me. I know from experience.<|endoftext|> [end of text]

```
llama_print_timings:        load time = 53526.48 ms
llama_print_timings:      sample time =   749.78 ms /   428 runs   (    1.75 ms per token,   570.83 tokens per second)
llama_print_timings: prompt eval time =  4232.80 ms /    10 tokens (  423.28 ms per token,     2.36 tokens per second)
llama_print_timings:        eval time = 532203.03 ms /   427 runs   ( 1246.38 ms per token,     0.80 tokens per second)
llama_print_timings:       total time = 537415.52 ms
```

Green-Sky commented:
@logicchains very nice, can you open a pr? even if its dirty, then mark it as a draft. :)

logicchains commented:
Done: https://github.com/ggerganov/llama.cpp/pull/3049 . It works as is if we're fine with having a lot of duplication between `convert-falcon180-hf-to-gguf.py` and `convert-falcon-hf-to-gguf.py`.

sekiis commented:
Thank you very much @logicchains the script convert works great on falcon180b.

I've encountered a situation where things aren't going well. I performed fine-tuning with a transformer using the "PEFT LORA" method, merged the base model with the LORA   "model.merge_and_unload() .. model.save_pretrained()"        
as save the output.
 I tested the fine-tuning through inference with the transformer framework, and everything was working perfectly up to this point my fine tuning work as expected with good response.

Now, I want to convert it into a q4 ggml, and during the "convert-falcon180...py" step, I noticed that there is a missing tensor: "lm_head.weight" -> output.weight. Despite this, both the convert.py and ./quantize steps proceed without errors. However, when I execute "./main ...," the tensor is missing, and the program returns an error: "error loading model: create_tensor: tensor 'output.weight' not found."

I'm currently trying to retrieve the missing tensor from the base model and add it to my fine-tuned model in the "convert-falcon180-hf-to-gguf.py" file, but I'm not sure if this is the correct approach.

Does anyone have any idea why the transformer is removing the "lm_head.weight" (output.weight) tensor and how to address this issue ?
